File: EVM.dart
import 'dart:io';
enum party {pti, ppp, pmln, mqm, independent}
int? voterCNIC;
bool isVoter = true;
List totalVoters = [];
List candidate = [
  {"name": "Imran Khan", "party": party.pti},
  {"name": "Bilawal Bhutto", "party": party.ppp},
  {"name": "Nawaz Sharif", "party": party.pmln},
  {"name": "Altaf Hussain", "party": party.mqm},
  {"name": "Independent", "party": party.independent,}];
List poolingAgent = [
  {"cnic" : 4220111569725 ,"username": "suhaib", "password": "suhaib", "party":party.mqm},
  {"cnic" : 4220111569726 ,"username": "sumair", "password": "sumair", "party": party.pti},
   {"cnic" : 4220111569727 ,"username": "asim", "password": "asim", "party": party.pti},
   {"cnic" : 4220111569728 ,"username": "arqum", "password": "arqum", "party": party.pmln},
];
List cnicNumb = [4220111569725,4220111569725,4220111569725,];
List ptiVoter = [42201, 42301];
List pmlnVoter = [];
List pppVoter = [1, 3, 23, 4, 2, 32];
List mqmVoter = [];
List independentVoter = [];
bool isloginCEC = false;
bool isRunning = true;
bool CECCanidateWindow = false;
bool CECPAWindow = false;
bool CECVoterWindow = false;
bool isLoginPA = false;

void main() {
  run();
}

run() {
  while (isRunning) {
    print("============ ELECTRONIC VOTING SYSTEM ============");
    print("Login as Cheif Election Commission. Press 1");
    print("Login as a Polling Agent. Press 2");
    print("to Cast your Vote. Press 3");
    print("Press Any Key to Stop Program");

    //first input to determine role.
    var input1 = stdin.readLineSync();
    if (input1 == "1") {
      cecFunctions();
    } else if (input1 == "2") {
      pollingAgentFunctions();
    } else if (input1 == "3") {
      VoterFunctions();
    } else {
      print("============Program Endeed!============");

      exit(0);
    }
  }
}

cecFunctions() {
  stdout.write("Enter Username: ");
  var inputUser = stdin.readLineSync();
  stdout.write("Enter Password: ");
  var inputPass = stdin.readLineSync();
  if (inputUser == "admin" && inputPass == "admin") {
    isloginCEC = true;
    print("============Login Successfull============");

    while (isloginCEC) {
      print("Press 1 for Canidates Menu.");
      print("Press 2 for  Polling Agents Menu.");
      print("Press 3 for Voters Menu.");
      print("Press Any Key to Main Menu.");
      var cecInput = stdin.readLineSync();
      if (cecInput == "1") {
        CECCanidateWindow = true;
        while (CECCanidateWindow) {
          print("Press 1 to add a Canidate.");
          print("Press 2 to check Canidate List.");
          print("Press 3 to update Canidate Details.");
          print("Press 4 to delete Canidate Details.");
          print("Press 0 to back to the previous Window.");

          var cecInputForCanidate = stdin.readLineSync();
          if (cecInputForCanidate == "1") {
            addCanidate();
          } else if (cecInputForCanidate == "2") {
            viewCanidate();
          } else if (cecInputForCanidate == "3") {
            updateCanidate();
          } else if (cecInputForCanidate == "4") {
            deleteCanidate();
          } else {
            CECCanidateWindow = false;
          }
        }
      } else if (cecInput == "2") {
        CECPAWindow = true;
        while (CECPAWindow) {
          print("Press 1 to add a Polling Agent.");
          print("Press 2 to check Polling Agent List.");
          print("Press 3 to update Polling Agent Details.");
          print("Press 4 to delete Polling Agent Details.");
          print("Press 0 to back to the previous Window.");

          var cecInputForPA = stdin.readLineSync();
          if (cecInputForPA == "1") {
            addPA();
          } else if (cecInputForPA == "2") {
            viewPA();
          } else if (cecInputForPA == "3") {
            updatePA();
          } else if (cecInputForPA == "4") {
            deletePA();
          } else {
            CECPAWindow = false;
          }
        }
      } else if (cecInput == "3") {
        CECVoterWindow = true;
        while (CECVoterWindow) {
          print("Press 1 to add a Voter Details.");
          print("Press 2 to check Voter List.");
          print("Press 3 to update Voter Details.");
          print("Press 4 to delete Voter Details.");
          print("Press 0 to back to the previous Window.");

          var cecInputForVoter = stdin.readLineSync();
          if (cecInputForVoter == "1") {
            addVoter();
          } else if (cecInputForVoter == "2") {
            viewVoter();
          } else if (cecInputForVoter == "3") {
            updateVoter();
          } else if (cecInputForVoter == "4") {
            deleteVoter();
          } else {
            CECVoterWindow = false;
          }
        }
      } else {
        print("Logout  Successfully.");
        isloginCEC = false;
      }
    }
  }else{
    print("Invalid Credentials. Please try again.");
  }
}

addCanidate() {
  print("Enter a Candidate name: ");
  var candidateInput = stdin.readLineSync();
  print("Please select a party for $candidateInput:");

  for (var i = 0; i < party.values.length; i++) {
    print("${i + 1}. ${party.values[i].name.toUpperCase()}");
  }

  var partyInput = stdin.readLineSync();
  int? selectedPartyIndex = int.tryParse(partyInput ?? '');
  if (selectedPartyIndex == null || selectedPartyIndex < 1 || selectedPartyIndex > party.values.length) {
    print("Invalid selection. Please try again.");
    return;
  }

  candidate.add({
    'name': candidateInput,
    'party': party.values[selectedPartyIndex - 1], // Save enum value directly
  });
  print("$candidateInput Value is Added Succesfully");
  print("========================================");
}

viewCanidate() {
  print("Candidates List: $candidate");
  // stdout.write(candidate);
}

void updateCanidate() {
  if (candidate.isEmpty) {
    print("No candidates available to update.");
    return;
  }

  print("Available Candidates:");
  for (var i = 0; i < candidate.length; i++) {
    print("Index $i: Name: ${candidate[i]["name"]}, Party: ${candidate[i]["party"]}");
  }

  print("Enter the index of the candidate to update:");
  var inputIndex = stdin.readLineSync();
  int? index = int.tryParse(inputIndex ?? '');

  if (index == null || index < 0 || index >= candidate.length) {
    print("Invalid index. Please try again.");
    return;
  }

  print("1. Update Name");
  print("2. Update Party");
  var choice = stdin.readLineSync();
  if (choice == "1") {
    print("Enter new name:");
    var newName = stdin.readLineSync();
    if (newName != null && newName.isNotEmpty) {
      candidate[index]["name"] = newName;
      print("Candidate name updated successfully.");
    } else {
      print("Invalid name. Update canceled.");
    }
  } else if (choice == "2") {
    print("Please select a new party:");
    for (var i = 0; i < party.values.length; i++) {
      print("${i + 1}. ${party.values[i].name.toUpperCase()}");
    }

    var partyInput = stdin.readLineSync();
    int? selectedPartyIndex = int.tryParse(partyInput ?? '');
    if (selectedPartyIndex == null || selectedPartyIndex < 1 || selectedPartyIndex > party.values.length) {
      print("Invalid selection. Please try again.");
      return;
    }

    candidate[index]["party"] = party.values[selectedPartyIndex - 1];
    print("Candidate party updated successfully.");
  } else {
    print("Invalid choice. Update canceled.");
  }
}
deleteCanidate() {
  for (var i = 0; i < candidate.length; i++) {
    print("Index: $i and Name is ${candidate[i]}");
  }
  print("Enter Index to Delete Candidate : ");
  int candidateIndex = int.parse(stdin.readLineSync()!);
  if(candidateIndex < 0 || candidateIndex >= candidate.length){
    print("Invalid index. Please try again.");
    return;
  }else{
  candidate.removeAt(candidateIndex);

  print("$candidateIndex Index Value is Deleted Succesfully");
  print("==================================================");
  }
}
void addPA() {
  print("Please Enter the CNIC Number of Polling Agent without dash: ");
  var cnicInput = stdin.readLineSync();
  if (cnicInput == null || cnicInput.length != 13 || int.tryParse(cnicInput) == null) {
    print("Invalid CNIC number. Please ensure it is 13 digits and numeric.");
    return;
  }

  int cnic = int.parse(cnicInput);
  if (cnicNumb.contains(cnic)) {
    print("CNIC Number already exists. Please try again.");
    return;
  }

  print("Enter Your Username to Add: ");
  var username = stdin.readLineSync();
  print("Enter Your Password: ");
  var password = stdin.readLineSync();

  print("Please select a party:");
  for (var i = 0; i < party.values.length; i++) {
    print("${i + 1}. ${party.values[i].name.toUpperCase()}");
  }

  var partyInput = stdin.readLineSync();
  int? selectedPartyIndex = int.tryParse(partyInput ?? '');
  if (selectedPartyIndex == null || selectedPartyIndex < 1 || selectedPartyIndex > party.values.length) {
    print("Invalid selection. Please try again.");
    return;
  }

  poolingAgent.add({
    "cnic": cnic,
    "username": username,
    "password": password,
    "party": party.values[selectedPartyIndex - 1]
  });
  cnicNumb.add(cnic); // Ensure CNIC is added to the global list.
  print("Polling Agent added successfully!");
}
viewPA() {
  print(poolingAgent);
  print("==================================================");
}


void updatePA() {
  for (var i = 0; i < poolingAgent.length; i++) {
    print("Index: $i and value is ${poolingAgent[i]}");
  }
  print("Please enter the index number to update cnic username ,password and party :");
  int inputPool = int.parse(stdin.readLineSync()!);
  if(inputPool < 0 || inputPool >= poolingAgent.length){
    print("Invalid index. Please try again.");
    return;
  }
  print("What would you like to update?");
  print("1: CNIC");
  print("2: Username");
  print("3: Password");
  print("4: Party");
  
  int fieldChoice = int.parse(stdin.readLineSync()!);

  switch (fieldChoice) {
    case 1:
      print("Please enter the new CNIC: ");
      var newCnicValue = stdin.readLineSync();
      poolingAgent[inputPool]["cnic"] = newCnicValue;
      print("CNIC updated successfully for index $inputPool.");
      break;

    case 2:
      print("Please enter the new username: ");
      var newUserNameValue = stdin.readLineSync();
      poolingAgent[inputPool]["username"] = newUserNameValue;
      print("Username updated successfully for index $inputPool.");
      break;

    case 3:
      print("Please enter the new password: ");
      var newPasswordValue = stdin.readLineSync();
      poolingAgent[inputPool]["password"] = newPasswordValue;
      print("Password updated successfully for index $inputPool.");
      break;

    case 4:
      print("Please select the new party (choose a number):");
      print("1: PTI");
      print("2: PPP");
      print("3: PMLN");
      print("4: MQM");
      print("5: Independent");
      int partyChoice = int.parse(stdin.readLineSync()!);

      party selectedParty;
      switch (partyChoice) {
        case 1:
          selectedParty = party.pti;
          break;
        case 2:
          selectedParty = party.ppp;
          break;
        case 3:
          selectedParty = party.pmln;
          break;
        case 4:
          selectedParty = party.mqm;
          break;
        case 5:
          selectedParty = party.independent;
          break;
        default:
          print("Invalid choice. Setting default party to Independent.");
          selectedParty = party.independent;
      }

      poolingAgent[inputPool]["party"] = selectedParty.toString().split('.').last;
      print("Party updated successfully to ${selectedParty.toString().split('.').last} for index $inputPool.");
      break;

    default:
      print("Invalid choice. No update made.");
      break;
  }

  print("==================================================");
}
deletePA() {
  for (var i = 0; i < poolingAgent.length; i++) {
    print("Index: $i and value is ${poolingAgent[i]}");
  }
  print("please enter index number to delete username and password");
  int inputPool = int.parse(stdin.readLineSync()!);
  if(inputPool < 0 || inputPool >= poolingAgent.length){
    print("Invalid index. Please try again.");
    return;
  }else{
  poolingAgent.removeAt(inputPool);
  print("$inputPool index value Deleted Succesfully");
  print("==================================================");
  }
}

addVoter() {
  print("Enter Cnic Number of Voter without dash:");
  int cnicInputs = int.parse(stdin.readLineSync()!);
 if(cnicInputs.toString().length != 13){
  if(cnicNumb.contains(cnicInputs)){
    print("CNIC Number already exists. Please try again.");
    }else{
  cnicNumb.add(cnicInputs);
  print("$cnicInputs Value is Added Succesfully");
  print("====================================");
  }
  }else{
    print("Invalid CNIC Number. Please try again.");
  }
}

viewVoter() {
  print("List of Voters.");
  print(cnicNumb);
  print("==================================================");
}

updateVoter() {
  for (var i = 0; i < cnicNumb.length; i++) {
    print("Index: ${i} Value: ${cnicNumb[i]}");
  }
  print("Enter Index Number for update.");
  int indexNumb = int.parse(stdin.readLineSync()!);
  if(indexNumb < 0 || indexNumb >= cnicNumb.length){
    print("Invalid index. Please try again.");
    return;
  }
  print("Enter Updated Cnic Number to $indexNumb index number");
  var updatedCnic = stdin.readLineSync();

  cnicNumb[indexNumb] = updatedCnic;
  print("$indexNumb Index Value is Updated Succesfully");
  print("=============================================");
}

deleteVoter() {
  for (var i = 0; i < cnicNumb.length; i++) {
    print("Index: ${i} Value: ${cnicNumb[i]}");
  }
  print("Enter Index Number To Delete Value: ");
  int indexNumb = int.parse(stdin.readLineSync()!);
  if(indexNumb < 0 || indexNumb >= cnicNumb.length){
    print("Invalid index. Please try again.");
    return;
  }else{
  cnicNumb.removeAt(indexNumb);
  print("$indexNumb Index Value is Deleted Succesfully");
  print("============================================");
  }
}

pollingAgentFunctions() {
  stdout.write("Enter Username: ");
  var inputUser = stdin.readLineSync();
  stdout.write("Enter Password: ");
  var inputPass = stdin.readLineSync();
  
  // Check login credentials
  var authenticatedAgent = poolingAgent.firstWhere(
    (pa) => pa["username"] == inputUser && pa["password"] == inputPass,
    orElse: () => null,
  );

  if (authenticatedAgent != null) {
    isLoginPA = true;
    var party = authenticatedAgent["party"];
    switch (party) {
      case "pti":
        totalVoters = ptiVoter;
        break;
      case "ppp":
        totalVoters = pppVoter;
        break;
      case "pmln":
        totalVoters = pmlnVoter;
        break;
      case "mqm":
        totalVoters = mqmVoter;
        break;
      case "independent":
        totalVoters = independentVoter;
        break;
      default:
        totalVoters = [];
    }

    print("============ Login Successful ============");
    print("Welcome, Polling Agent for $party!");

    // Polling Agent Menu
    while (isLoginPA) {
      print("\nPolling Agent Menu:");
      print("1. View total votes gained by your party.");
      print("2. Log out.");
      print("=========================================");

      var agentInput = stdin.readLineSync();
      if (agentInput == "1") {
        print("Total Votes for $party: ${totalVoters.length}");
      } else if (agentInput == "2") {
        isLoginPA = false;
        print("You have been logged out successfully.");
      } else {
        print("Invalid option. Please try again.");
      }
    }
  } else {
    print("Invalid credentials. Please try again.");
  }
}
VoterFunctions() {
  stdout.write("Enter your CNIC Number: ");
  int cnic = int.parse(stdin.readLineSync()!);
  
  if (cnicNumb.contains(cnic)) {
    if (ptiVoter.contains(cnic) || pppVoter.contains(cnic) || pmlnVoter.contains(cnic) || mqmVoter.contains(cnic) || independentVoter.contains(cnic)) {
      print("You have already voted!");
    } else {
      print("Vote for a Party:");
      print("1. PTI");
      print("2. PPP");
      print("3. PML-N");
      print("4. MQM");
      print("5. Independent");
      var voteInput = stdin.readLineSync();
      
      if (voteInput == "1") {
        ptiVoter.add(cnic);
        print("Your vote for PTI has been cast successfully.");
      } else if (voteInput == "2") {
        pppVoter.add(cnic);
        print("Your vote for PPP has been cast successfully.");
      } else if (voteInput == "3") {
        pmlnVoter.add(cnic);
        print("Your vote for PML-N has been cast successfully.");
      } else if (voteInput == "4") {
        mqmVoter.add(cnic);
        print("Your vote for MQM has been cast successfully.");
      } else if (voteInput == "5") {
        independentVoter.add(cnic);
        print("Your vote for Independent has been cast successfully.");
      } else {
        print("Invalid Input. Vote not cast.");
      }
    }
  } else {
    print("CNIC not found in the Voter List. Please register first.");
  }
}partyVotes() {
  print("Your Party Total Votes: ");
  if (totalVoters == ptiVoter) {
    print(ptiVoter.length);
  } else if (totalVoters == pppVoter) {
    print(pppVoter.length);
  } else if (totalVoters == pmlnVoter) {
    print(pmlnVoter.length);
  }
  print("==================================================");
}

totalVotes() {
  print("Total votes of all party.");

  int list1Length = ptiVoter.length;
  int list2Length = pppVoter.length;
  int list3Length = pmlnVoter.length;
  int list4Length = mqmVoter.length;
  int list5Length = independentVoter.length;

  List<int> mergedList = [];
  mergedList = [
    ...ptiVoter,
    ...pppVoter,
    ...pmlnVoter,
    ...mqmVoter,
    ...independentVoter
  ];

  // Print the length of each list and the merged list.
  print("The Voter of PTI is ${list1Length}.");
  print("The Voter of PPP is ${list2Length}.");
  print("The Voter of PMLN is ${list3Length}.");
  print("The Voter of MQM is ${list4Length}.");
  print("The Voter of INDEPENDENT is ${list5Length}.");
  print("The Total Voter of ALL Party is ${mergedList.length}.");

  print("==================================================");
}

highestvotes() {
  int max = findMaxListLength(
      [ptiVoter, pmlnVoter, pppVoter, mqmVoter, independentVoter]);
  print("The highest vote value is: $max");
}

int findMaxListLength(List<List> lists) {
  int maxLength = 0;

  for (List list in lists) {
    if (list.length > maxLength) {
      maxLength = list.length;
    }
  }

  return maxLength;
}

File: evm.py
import sys

class Party:
    PTI = "pti"
    PPP = "ppp"
    PMLN = "pmln"
    MQM = "mqm"
    INDEPENDENT = "independent"

is_voter = True
total_voters = []
candidates = [
    {"name": "Imran Khan", "party": Party.PTI},
    {"name": "Bilawal Bhutto", "party": Party.PPP},
    {"name": "Nawaz Sharif", "party": Party.PMLN},
    {"name": "Altaf Hussain", "party": Party.MQM},
    {"name": "Independent", "party": Party.INDEPENDENT},
]
polling_agents = [
    {"cnic": 4220111569725, "username": "suhaib", "password": "suhaib", "party": Party.MQM},
    {"cnic": 4220111569726, "username": "sumair", "password": "sumair", "party": Party.PTI},
    {"cnic": 4220111569727, "username": "asim", "password": "asim", "party": Party.PTI},
    {"cnic": 4220111569728, "username": "arqum", "password": "arqum", "party": Party.PMLN},
]
# cnic_numbers = [4220111569725, 4220111569725, 4220111569725]
pti_voters = [422014, 42301]
pmln_voters = []
ppp_voters = [1, 3, 23, 4, 2, 32]
mqm_voters = []
independent_voters = []
is_login_cec = False
is_running = True
cec_candidate_window = False
cec_pa_window = False
cec_voter_window = False
is_login_pa = False

def main():
    run()

def run():
    global is_running
    while is_running:
        print("============ ELECTRONIC VOTING SYSTEM ============")
        print("Login as Chief Election Commission. Press 1")
        print("Login as a Polling Agent. Press 2")
        print("To cast your vote. Press 3")
        print("Press Any Key to Stop Program")

        input1 = input()
        if input1 == "1":
            cec_functions()
        elif input1 == "2":
            polling_agent_functions()
        elif input1 == "3":
            voter_functions()
        else:
            print("============ Program Ended! ============")
            sys.exit(0)
def polling_agent_functions():
    print("============ Polling Agent Login ============")
    input_user = input("Enter Username: ")
    input_pass = input("Enter Password: ")

    # Check login credentials
    authenticated_agent = next(
        (pa for pa in polling_agents if pa["username"] == input_user and pa["password"] == input_pass),
        None
    )

    if authenticated_agent:
        global is_login_pa, total_voters
        is_login_pa = True
        party = authenticated_agent["party"]
        if party == "pti":
            total_voters = pti_voters
        elif party == "ppp":
            total_voters = ppp_voters
        elif party == "pmln":
            total_voters = pmln_voters
        elif party == "mqm":
            total_voters = mqm_voters
        elif party == "independent":
            total_voters = independent_voters
        else:
            total_voters = []

        print("============ Login Successful ============")
        print(f"Welcome, Polling Agent for {party}!")

        # Polling Agent Menu
        while is_login_pa:
            print("\nPolling Agent Menu:")
            print("1. View total votes gained by your party.")
            print("2. Log out.")
            print("=========================================")

            agent_input = input("Enter your choice: ")
            if agent_input == "1":
                print(f"Total Votes for {party}: {len(total_voters)}")
            elif agent_input == "2":
                is_login_pa = False
                print("You have been logged out successfully.")
            else:
                print("Invalid option. Please try again.")
    else:
        print("Invalid credentials. Please try again.")


def voter_functions():
    print("============ Voter Login ============")
    try:
        cnic = int(input("Enter your CNIC Number: "))
        
        # Check if the CNIC exists in the voter list (by comparing with 'cnic' key in dictionaries)
        if any(voter['cnic'] == cnic for voter in voters):
            if any(cnic in voter_list for voter_list in [pti_voters, ppp_voters, pmln_voters, mqm_voters, independent_voters]):
                print("You have already voted!")
            else:
                print("Vote for a Party:")
                print("1. PTI")
                print("2. PPP")
                print("3. PML-N")
                print("4. MQM")
                print("5. Independent")
                vote_input = input("Enter your choice: ")

                if vote_input == "1":
                    pti_voters.append(cnic)
                    print("Your vote for PTI has been cast successfully.")
                elif vote_input == "2":
                    ppp_voters.append(cnic)
                    print("Your vote for PPP has been cast successfully.")
                elif vote_input == "3":
                    pmln_voters.append(cnic)
                    print("Your vote for PML-N has been cast successfully.")
                elif vote_input == "4":
                    mqm_voters.append(cnic)
                    print("Your vote for MQM has been cast successfully.")
                elif vote_input == "5":
                    independent_voters.append(cnic)
                    print("Your vote for Independent has been cast successfully.")
                else:
                    print("Invalid Input. Vote not cast.")
        else:
            print("CNIC not found in the Voter List. Please register first.")
    except ValueError:
        print("Invalid CNIC. Please enter a valid number.")
def cec_functions():
    global is_login_cec, cec_candidate_window, cec_pa_window, cec_voter_window
    username = input("Enter Username: ")
    password = input("Enter Password: ")
    if username == "admin" and password == "admin":
        is_login_cec = True
        print("============ Login Successful ============")

        while is_login_cec:
            print("Press 1 for Candidates Menu.")
            print("Press 2 for Polling Agents Menu.")
            print("Press 3 for Voters Menu.")
            print("Press Any Key to Main Menu.")
            cec_input = input()

            if cec_input == "1":
                cec_candidate_window = True
                while cec_candidate_window:
                    print("Press 1 to add a Candidate.")
                    print("Press 2 to check Candidate List.")
                    print("Press 3 to update Candidate Details.")
                    print("Press 4 to delete Candidate Details.")
                    print("Press 0 to back to the previous Window.")

                    candidate_input = input()
                    if candidate_input == "1":
                        add_candidate()
                    elif candidate_input == "2":
                        view_candidates()
                    elif candidate_input == "3":
                        update_candidate()
                    elif candidate_input == "4":
                        delete_candidate()
                    else:
                        cec_candidate_window = False
            elif cec_input == "2":
                cec_pa_window = True
                while cec_pa_window:
                    print("Press 1 to add a Polling Agent.")
                    print("Press 2 to check Polling Agent List.")
                    print("Press 3 to update Polling Agent Details.")
                    print("Press 4 to delete Polling Agent Details.")
                    print("Press 0 to back to the previous Window.")

                    pa_input = input()
                    if pa_input == "1":
                        add_polling_agent()
                    elif pa_input == "2":
                        view_polling_agents()
                    elif pa_input == "3":
                        update_polling_agent()
                    elif pa_input == "4":
                        delete_polling_agent()
                    else:
                        cec_pa_window = False
            elif cec_input == "3":
                cec_voter_window = True
                while cec_voter_window:
                    print("Press 1 to add Voter Details.")
                    print("Press 2 to check Voter List.")
                    print("Press 3 to update Voter Details.")
                    print("Press 4 to delete Voter Details.")
                    print("Press 0 to back to the previous Window.")

                    voter_input = input()
                    if voter_input == "1":
                        add_voter()
                    elif voter_input == "2":
                        view_voters()
                    elif voter_input == "3":
                        update_voter()
                    elif voter_input == "4":
                        delete_voter()
                    else:
                        cec_voter_window = False
            else:
                print("Logout Successfully.")
                is_login_cec = False
    else:
        print("Invalid Credentials. Please try again.")

def add_candidate():
    name = input("Enter a Candidate name: ")
    print(f"Please select a party for {name}:")
    parties = [Party.PTI, Party.PPP, Party.PMLN, Party.MQM, Party.INDEPENDENT]
    for idx, party in enumerate(parties, start=1):
        print(f"{idx}. {party.upper()}")

    try:
        party_choice = int(input())
        selected_party = parties[party_choice - 1]
    except (ValueError, IndexError):
        print("Invalid selection. Please try again.")
        return

    candidates.append({"name": name, "party": selected_party})
    print(f"{name} is Added Successfully.")

def view_candidates():
    print("Candidates List:")
    for candidate in candidates:
        print(candidate)

def update_candidate():
    print("Update Candidate Details:")
    for idx, candidate in enumerate(candidates, start=1):
        print(f"{idx}. {candidate['name']} ({candidate['party']})")

    try:
        choice = int(input("Select the candidate to update (by number): "))
        if 1 <= choice <= len(candidates):
            selected_candidate = candidates[choice - 1]
            new_name = input(f"Enter new name for {selected_candidate['name']} (leave blank to keep unchanged): ")
            print("Please select a new party:")
            parties = [Party.PTI, Party.PPP, Party.PMLN, Party.MQM, Party.INDEPENDENT]
            for idx, party in enumerate(parties, start=1):
                print(f"{idx}. {party.upper()}")

            try:
                party_choice = int(input())
                selected_party = parties[party_choice - 1]
            except (ValueError, IndexError):
                print("Invalid selection. Keeping the old party.")
                selected_party = selected_candidate['party']

            if new_name:
                selected_candidate['name'] = new_name
            selected_candidate['party'] = selected_party
            print("Candidate details updated successfully.")
        else:
            print("Invalid selection. Please try again.")
    except ValueError:
        print("Invalid input. Please try again.")

def delete_candidate():
    print("Delete Candidate:")
    for idx, candidate in enumerate(candidates, start=1):
        print(f"{idx}. {candidate['name']} ({candidate['party']})")

    try:
        choice = int(input("Select the candidate to delete (by number): "))
        if 1 <= choice <= len(candidates):
            removed_candidate = candidates.pop(choice - 1)
            print(f"{removed_candidate['name']} has been removed.")
        else:
            print("Invalid selection. Please try again.")
    except ValueError:
        print("Invalid input. Please try again.")

def add_polling_agent():
    try:
        cnic = int(input("Enter CNIC for the polling agent: "))
        username = input("Enter username for the polling agent: ")
        password = input("Enter password for the polling agent: ")
        print("Please select a party for the polling agent:")
        parties = [Party.PTI, Party.PPP, Party.PMLN, Party.MQM, Party.INDEPENDENT]
        for idx, party in enumerate(parties, start=1):
            print(f"{idx}. {party.upper()}")

        try:
            party_choice = int(input())
            selected_party = parties[party_choice - 1]
        except (ValueError, IndexError):
            print("Invalid selection. Please try again.")
            return

        polling_agents.append({
            "cnic": cnic,
            "username": username,
            "password": password,
            "party": selected_party
        })
        print("Polling agent added successfully.")
    except ValueError:
        print("Invalid CNIC. Please try again.")

def view_polling_agents():
    print("Polling Agents List:")
    for agent in polling_agents:
        print(f"CNIC: {agent['cnic']}, Username: {agent['username']}, Party: {agent['party']}")

def update_polling_agent():
    print("Update Polling Agent Details:")
    for idx, agent in enumerate(polling_agents, start=1):
        print(f"{idx}. {agent['username']} (CNIC: {agent['cnic']}, Party: {agent['party']})")

    try:
        choice = int(input("Select the polling agent to update (by number): "))
        if 1 <= choice <= len(polling_agents):
            selected_agent = polling_agents[choice - 1]
            new_username = input(f"Enter new username for {selected_agent['username']} (leave blank to keep unchanged): ")
            new_password = input(f"Enter new password (leave blank to keep unchanged): ")
            print("Please select a new party:")
            parties = [Party.PTI, Party.PPP, Party.PMLN, Party.MQM, Party.INDEPENDENT]
            for idx, party in enumerate(parties, start=1):
                print(f"{idx}. {party.upper()}")

            try:
                party_choice = int(input())
                selected_party = parties[party_choice - 1]
            except (ValueError, IndexError):
                print("Invalid selection. Keeping the old party.")
                selected_party = selected_agent['party']

            if new_username:
                selected_agent['username'] = new_username
            if new_password:
                selected_agent['password'] = new_password
            selected_agent['party'] = selected_party
            print("Polling agent details updated successfully.")
        else:
            print("Invalid selection. Please try again.")
    except ValueError:
        print("Invalid input. Please try again.")

def delete_polling_agent():
    print("Delete Polling Agent:")
    for idx, agent in enumerate(polling_agents, start=1):
        print(f"{idx}. {agent['username']} (CNIC: {agent['cnic']}, Party: {agent['party']})")

    try:
        choice = int(input("Select the polling agent to delete (by number): "))
        if 1 <= choice <= len(polling_agents):
            removed_agent = polling_agents.pop(choice - 1)
            print(f"{removed_agent['username']} has been removed.")
        else:
            print("Invalid selection. Please try again.")
    except ValueError:
        print("Invalid input. Please try again.")

# List to store voter details
voters = [{'cnic': 42201, 'name': 'suhaib', 'age': 22, 'address': 'malir', 'polling_station': 'malir'}]  # Format: [{"cnic": int, "name": str, "age": int, "address": str, "polling_station": str}]

def add_voter():
    try:
        cnic = int(input("Enter CNIC for the voter: "))
        name = input("Enter name of the voter: ")
        age = int(input("Enter age of the voter: "))
        address = input("Enter address of the voter: ")
        polling_station = input("Enter polling station for the voter: ")

        voters.append({
            "cnic": cnic,
            "name": name,
            "age": age,
            "address": address,
            "polling_station": polling_station
        })
        print("Voter added successfully.")
    except ValueError:
        print("Invalid input. Please ensure CNIC and age are numeric values.")

def view_voters():
    print("List of Registered Voters:")
    if not voters:
        print("No voters registered yet.")
        return

    for idx, voter in enumerate(voters, start=1):
        print(f"{idx}. CNIC: {voter['cnic']}, Name: {voter['name']}, Age: {voter['age']}, "
              f"Address: {voter['address']}, Polling Station: {voter['polling_station']}")

def update_voter():
    print("Update Voter Details:")
    if not voters:
        print("No voters to update.")
        return

    for idx, voter in enumerate(voters, start=1):
        print(f"{idx}. {voter['name']} (CNIC: {voter['cnic']}, Polling Station: {voter['polling_station']})")

    try:
        choice = int(input("Select the voter to update (by number): "))
        if 1 <= choice <= len(voters):
            selected_voter = voters[choice - 1]
            new_name = input(f"Enter new name for {selected_voter['name']} (leave blank to keep unchanged): ")
            try:
                new_age = input(f"Enter new age for {selected_voter['name']} (leave blank to keep unchanged): ")
                new_age = int(new_age) if new_age else selected_voter['age']
            except ValueError:
                print("Invalid input for age. Keeping old value.")
                new_age = selected_voter['age']

            new_address = input(f"Enter new address (leave blank to keep unchanged): ")
            new_polling_station = input(f"Enter new polling station (leave blank to keep unchanged): ")

            if new_name:
                selected_voter['name'] = new_name
            selected_voter['age'] = new_age
            if new_address:
                selected_voter['address'] = new_address
            if new_polling_station:
                selected_voter['polling_station'] = new_polling_station

            print("Voter details updated successfully.")
        else:
            print("Invalid selection. Please try again.")
    except ValueError:
        print("Invalid input. Please try again.")

def delete_voter():
    print("Delete Voter:")
    if not voters:
        print("No voters to delete.")
        return

    for idx, voter in enumerate(voters, start=1):
        print(f"{idx}. {voter['name']} (CNIC: {voter['cnic']})")

    try:
        choice = int(input("Select the voter to delete (by number): "))
        if 1 <= choice <= len(voters):
            removed_voter = voters.pop(choice - 1)
            print(f"{removed_voter['name']} has been removed.")
        else:
            print("Invalid selection. Please try again.")
    except ValueError:
        print("Invalid input. Please try again.")

if __name__ == "__main__":
    main()
